misc/codegen/test/test_render.py

```diff
@@ -76,6 +76,45 @@ def test_render(pystache_renderer, sut):
     ]
 
 
+def test_render_provided_template(pystache_renderer, sut):
```

Comment:
Are these unit tests?  How should I read them?

Reply:
yes. These tests mock the underlying third-party library in charge of doing the actual template instantiation. That is the `pystache_rendeder` test fixture. In this line
```python
pystache_renderer.render_name.side_effect = (text,)
```
we are saying that when `render_name` is called the first time, it will return `text` (and fail any following call). In this line 
```python
    assert pystache_renderer.mock_calls == [
        mock.call.render_name(data.template, data, generator=generator),
    ]
```
we check that is was indeed called once with the provided arguments.

Then we check that the output file is indeed created and its contents is actually `text` (which is what we set as the return value of the internal template rendering). Thanks to mocking, we don't need to deal with any actual "real" data: `data` is actually an empty mock object, as we are only checking that it is passed as is to the underlying library call.